src/plots/cartesian/index.js

```diff
@@ -189,41 +195,60 @@ function plotOne(gd, plotinfo, cdSubplot, transitionOpts, makeOnCompleteCallback
     var modules = fullLayout._modules;
     var _module, cdModuleAndOthers, cdModule;
 
+    // Separate traces by zindex and plot each zindex group separately
+    // TODO: Performance
+    var traceZindexGroups = {};
+    for(var t = 0; t < cdSubplot.length; t++) {
+        var trace = cdSubplot[t][0].trace;
+        var zi = trace.zindex || 0;
+        if(!traceZindexGroups[zi]) traceZindexGroups[zi] = [];
+        traceZindexGroups[zi].push(cdSubplot[t]);
+    }
+
     var layerData = [];
     var zoomScaleQueryParts = [];
 
-    for(var i = 0; i < modules.length; i++) {
-        _module = modules[i];
-        var name = _module.name;
-        var categories = Registry.modules[name].categories;
-
-        if(categories.svg) {
-            var className = (_module.layerName || name + 'layer');
-            var plotMethod = _module.plot;
-
-            // plot all visible traces of this type on this subplot at once
-            cdModuleAndOthers = getModuleCalcData(cdSubplot, plotMethod);
-            cdModule = cdModuleAndOthers[0];
-            // don't need to search the found traces again - in fact we need to NOT
-            // so that if two modules share the same plotter we don't double-plot
-            cdSubplot = cdModuleAndOthers[1];
-
-            if(cdModule.length) {
-                layerData.push({
-                    i: traceLayerClasses.indexOf(className),
-                    className: className,
-                    plotMethod: plotMethod,
-                    cdModule: cdModule
-                });
-            }
+    // Plot each zindex group in ascending order
+    var zindices = Object.keys(traceZindexGroups)
+        .map(Number)
```

Comment:
Do we really need to convert to number here?

Reply:
We need to cast these indices to numbers since they are obtained from the keys. In `getModuleCalcData` we compare `trace.zindex === zindex` so they have to be `Numbers`.

We could also parse them in the comparison, but since it's inside of the loop that parsing will occur more times.